alternator/scoped_item_list.hh

```diff
@@ -0,0 +1,181 @@
+                for (auto&& gi : owner->_gentle_iterators) {
+                    if (gi.iter == iter) {
+                        gi.iter++;
+                    }
+                }
+                owner->_list.erase(iter);
+            }
+        }
+        T& get() { return value; }
+        T* operator->() { return &value; }
+        T& operator*() { return value; }
+
+        handle(handle&& other) noexcept :
+            boost::intrusive::list_base_hook<>(std::move(other)), 
+            value(std::move(other.value)), owner(other.owner) {
+                other.owner = nullptr;
+        }
+        handle& operator=(handle&& other) = delete;
+        handle(const handle&) = delete;
+        handle& operator=(const handle&) = delete;
+    };
+
+    using list_type = boost::intrusive::list<handle, boost::intrusive::constant_time_size<false>>;
+    list_type _list;
+
+public:
+    template <typename... Args>
+    handle emplace(Args&&... args) {
+        handle ret(this, std::forward<Args>(args)...);
+        _list.push_back(ret);
+        return ret;
+    }
+    scoped_item_list() = default;
+    ~scoped_item_list() {
+        // We usually expect that when the scoped_item_list<T> is destroyed,
+        // the list is already empty. For example, in the Alternator server,
+        // the list is destroyed when the server object is destroyed, and the
+        // server is only destroyed after all requests are done - so all
+        // handles those requests held to list items have been destroyed, so
+        // the list is empty.
+        // But if for some reason the list is not empty, it's not a disaster
+        // either - we just remove the "owner" pointer from all the handles.
+        // Each handle will continue to hold its item as before, but it is
+        // detached from any list.
+        for (auto& h : _list) {
+            h.owner = nullptr;
+        }
+        // We don't expect that the scoped_item_list<T> be destroyed while any
+        // for_each_gently() are still running. In general trying to do that
+        // doesn't make sense - since for_each_gently() is a method of the
+        // object being destroyed. We could have allowed this by making the
+        // _gentle_iterators list an intrusive list as well, but it's not
+        // worth the effort, since we don't expect this to ever happen.
+        if (!_gentle_iterators.empty()) {
+            utils::on_fatal_internal_error("scoped_item_list destroyed while "
+                "for_each_gently() is still running");
+        }
+    }
+    scoped_item_list(const scoped_item_list&) = delete;
+    scoped_item_list& operator=(const scoped_item_list&) = delete;
+
+    bool empty() const { return _list.empty(); }
+
+    struct gentle_iterator {
+        list_type::iterator iter, end;
+        gentle_iterator(scoped_item_list<T>& s) : iter(s._list.begin()), end(s._list.end()) {}
```

Comment:
Isn't it more accurate to sample end() each time (and capture s._list, not s.list.end())?

It ends up the same, because the iterator points at the list, but seems safer.

Reply:
Done (deleted from generic_server.hh). But I did NOT address your original comment yet - which was about the end() thing in the original code which I copied.